### ISSUE TITLE
Possible implementation mistake

### ISSUE THREAD

**Issue by AwesomeLemon** [Daniil-Osokin/fully-learnable-group-convolution.pytorch#4]
Hi.

I think there's a mistake in your implementation: doing binarization the way you do it doesn't allow gradients to flow through. This means that ``out_channels_in_group_assignment_map`` and ``in_channels_in_group_assignment_map`` are never updated. (Because autograd of Pytorch doesn't record ``==`` as ``grad_fn``). And in any case notice that in the paper binarized matrices are used only in forward step, not in the backward step.

However, I have only tested it on my own problem, so maybe I'm misunderstanding something. I'd recommend you to check whether values of ``in_channels_in_group_assignment_map`` change in your case. In case they don't, since I've managed to solve this problem in my code, I'd be glad to create a pull-request.

**Comment by Daniil-Osokin** [comment]
Nice catch :muscle:, thanks for pointing this out! Mind to check #5?

**Comment by AwesomeLemon** [comment]
I think the approach with extending autograd.Function is correct, but I'm not sure whether you should mask the gradient on the backward pass. Because if you do so, it means you only ever update the biggest value, and it seems quite unlikely that it will decrease; so with very high probability the filter will not change its group during the training

**Comment by Daniil-Osokin** [comment]
Thanks! Yes, masking only passes gradient to the biggest value. However there is `softmax`, which spreads it to all, so all values may change. Actually I did not catch why the biggest value will not decrease, can you please give a hint :)?

**Comment by AwesomeLemon** [comment]
After going through the math carefully I now see that you're completely correct and my intuition was wrong. Thanks for your responses & the quick fix. :)
It would be cool in any case if you'd check whether filters change their group assignment a lot: I think it's useful to know, and the authors don't mention it.

**Comment by Daniil-Osokin** [comment]
Yeah, that is an interesting thing to test (even because with the first buggy "random" groups result is similar by accuracy). Will do that.

**Comment by AwesomeLemon** [comment]
great, I'll be looking forward to hearing what you'll find

**Comment by Daniil-Osokin** [comment]
Please, check #7. In short, they hardly change initial random assignment, possibly still be a bug somewhere.

**Comment by Daniil-Osokin** [comment]
Hope, it is clear now.

**Comment by AwesomeLemon** [comment]
Sorry for not responding for a long time. I've checked #7 , thanks for investigation. Indeed, it looks as if  the performance of the Flgc algorithm is explained just by it having random sparse connectivity. Pity.

Although, it may also be that while the number of filter per group hardly changes, filters simply swap groups. So for example filters 1,2,3 were in group A, and filters 4,5,6 were in group B; after a single epoch suppose 1,2,6 were in group A, and 3,4,5 in group B. This would be a reasonably big change, and it wouldn't be visible in the testing script you've written (since number of filters in each group stayed the same)
